cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/includes/bottom.jsp

```diff
@@ -23,30 +23,45 @@
 
 </div>  <!-- END #content -->
 
+<spring:eval var="serverName" expression="@casProperties.getProperty('server.name')"/>
+<spring:eval var="osfLoginUrl" expression="@casProperties.getProperty('cas.osf.login.url')"/>
+<spring:eval var="casLogoutUrl" expression="@casProperties.getProperty('cas.logout.url')"/>
+<spring:eval var="createAccountUrl" expression="@casProperties.getProperty('osf.createAccount.url')"/>
+<spring:eval var="osfUrl" expression="@casProperties.getProperty('osf.url')"/>
+
 <c:if test="${empty serviceParam}">
-    <%-- Try to get the service from the login context first. If failed, try to get it from the request parameters. --%>
-    <c:set var="serviceParam" value="&service=${osfLoginContext.isServiceUrl() ? osfLoginContext.getServiceUrl() : param.service ? fn:escapeXml(param.service) : ''}"/>
+    <%-- Try to obtain the service from the login context first. If failed, get it from the request parameters. --%>
+    <c:set var="serviceUrl" value="${osfLoginContext.isServiceUrl() ? osfLoginContext.getServiceUrl() : fn:escapeXml(param.service)}"/>
+    <c:set var="serviceParam" value="&service=${serviceUrl}"/>
 </c:if>
 
 <div class="bottom-link">
     <br/>
+    <c:if test="${linkSignOutandBackToSignIn}">
+        <c:set var="osfFullLoginUrl" value="${serverName}${osfLoginUrl}" />
```

Comment:
From line 41 to 48, the "sign out and sign in with existing service" feature requires full URL with domain and URL encoding to work 100% all the time.